### PR TITLE
torch.mtia module for MTIA device backend

### PR DESCRIPTION
**Pull request by egienvalue**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123614
* __->__ #123612
* #123611

MTIA device has its own Module in PyTorch now.
torch.mtia has following APIs similar to other backends. The lazy_init is also supported.
```
__all__ = [
    "init",
    "is_available",
    "synchronize",
    "device_count",
    "current_device",
    "current_stream",
    "default_stream",
    "set_stream",
    "stream",
    "device",
]

```
------------
For device management. We expand AccleratorHooksInterface to support generic device management and it can be used in both C++ and PyThon.
```
def _accelerator_hooks_device_count() -> _int: ...
def _accelerator_hooks_set_current_device(device_index: _int) -> None: ...
def _accelerator_hooks_get_current_device() -> _int : ...
def _accelerator_hooks_exchange_device(device_index: _int) -> _int : ...
def _accelerator_hooks_maybe_exchange_device(device_index: _int) -> _int : ...
```

---------
Adding get_device_module API to retrieve device modules for different device types.
```
def get_device_module(device: Optional[Union[torch.device, str]] = None)
```
---------

### CODE REVIEW

File: aten/src/ATen/DeviceAccelerator.cpp
Comment (albanD) [inline aten/src/ATen/DeviceAccelerator.cpp:21]: Can you add a CHECK_NO_MTIA macro and add it to the other conditions?
Also add NO_CUDA/NO_PU1 to this branch

File: aten/src/ATen/detail/AcceleratorHooksInterface.h
Comment (albanD) [inline aten/src/ATen/detail/AcceleratorHooksInterface.h:19]: Why do you add an implementation here instead of keeping it pure virtual?

File: docs/source/mtia.rst
Comment (albanD) [inline docs/source/mtia.rst:1]: nit: can you add a bit more doc here to mention that this device is implemented our of tree and only the API is defined here.

File: torch/__init__.py
Comment (albanD) [inline torch/__init__.py:2016]: I think that for the general case we can follow the approach from https://github.com/pytorch/pytorch/blob/497bac223c72c14df85353512a77e60f3c8a6290/torch/_utils.py#L894
(we should also remove that private function and use this one in a follow up).

Also we should make sure that this function accepts any Device-like input and do appropriate typechecking.

File: torch/mtia/__init__.py
Comment (albanD) [inline torch/mtia/__init__.py:98]: The mtia/Module.cpp is always built given the current build config? So this will always be True?
tbh I prefer it this way (the same API is always available, you can have a `is_built` flag in c++ that is ifdef-ed to return True/False based on `USE_MTIA` build flag.

**Review by albanD — COMMENTED** [review]
The accelerator_hooks API looks a bit weird in this PR, but I think they make sense with the idea that we can have a base accelerator device module that uses them that is shared in the future.

File: aten/src/ATen/detail/AcceleratorHooksInterface.h
Comment (egienvalue) [reply aten/src/ATen/detail/AcceleratorHooksInterface.h]: Let me revert it back. It should be pure virtual if every device backend requires to implement it.

File: torch/__init__.py
Comment (albanD) [inline torch/__init__.py:2024]: This should be elif device is None ?
And a final else that is an error for invalid input type

File: torch/csrc/mtia/Module.cpp
Comment (albanD) [inline torch/csrc/mtia/Module.cpp:18]: Is MTIA driver not fork-safe? :p

**Comment by albanD** [comment]
The `get_device_module` comment is the only big outstanding thing. Will be good to go after this one.

File: torch/csrc/mtia/Module.cpp
Comment (egienvalue) [reply torch/csrc/mtia/Module.cpp:18]: We actually haven't verified it yet but I feel it doesn't support fork, the device management is different.

File: aten/src/ATen/detail/MTIAHooksInterface.cpp
Comment (egienvalue) [reply aten/src/ATen/detail/MTIAHooksInterface.cpp:13]: I see. Thanks for the info. Let me remove those C10_MOBILE stuff, I just saw other device backends have it. I thought it is a must have thing to guarantee the compatibility.

File: torch/__init__.py
Comment (egienvalue) [reply torch/__init__.py:2024]: Oh, let me address this part.

File: torch/__init__.py
Comment (albanD) [inline torch/__init__.py:2023]: Please specify here that if no device is given, this will return the module for the current accelerator or CPU if none is present

**Review by albanD — APPROVED** [review]
Small nit in the doc, good to go otherwise!

**Comment by egienvalue** [comment]
@egienvalue has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56443356).

**Comment by egienvalue** [comment]
> @pytorchbot revert -m "this PR broke ROCm with message RuntimeError: Cannot have MTIA with other devices" -c nosignal

@jeffdaily I think you reverted the wrong pr. And in the same time, I have already had a forward fix in #124924

**Comment by jeffdaily** [comment]
> > @pytorchbot revert -m "this PR broke ROCm with message RuntimeError: Cannot have MTIA with other devices" -c nosignal
> 
> @jeffdaily I think you reverted the wrong pr. And in the same time, I have already had a forward fix in #124924

We have a large number of failures on the HUD that we're trying to resolve.  I did not see the forward fix.  My apologies.  Please reland and tag me.  I'm not trying to be a pain, I'm just trying to help recover our ROCm CI signal.

**Comment by jeffdaily** [comment]
Perhaps when you go to reland this one, combine it with the changes in your forward fix PR.

**Review by albanD — CHANGES_REQUESTED** [review]
Please add the forward fix in here before merging it again. We definitely should not merge PRs that are known to be broken!

**Comment by egienvalue** [comment]
> Please add the forward fix in here before merging it again. We definitely should not merge PRs that are known to be broken!

Yea, I am working on it. BTW, I found ROCm CIs typically won't be triggered automatically. Is there easy way to manually trigger them ? @jeffdaily Adding ROCm label to the PR ?

**Review by albanD — APPROVED** [review]
Thanks for adding this back. Not sure what is the right way to trigger all the necessary rocm signal. Please wait for @jeffdaily input here before merging.

**Comment by jithunnair-amd** [comment]
> Thanks for adding this back. Not sure what is the right way to trigger all the necessary rocm signal. Please wait for @jeffdaily input here before merging.

Yes `ciflow/rocm` should be the extensive list of single-GPU unit tests for ROCm. Add `ciflow/inductor` (usually added automatically) for inductor-related PRs, and `ciflow/periodic` for distributed-logic related PRs.

**Comment by jeffdaily** [comment]
@egienvalue @albanD There are _some_ rocm jobs that run as part of ciflow/trunk before merge.  A subset of the full ciflow/rocm suite.  We don't have runner capacity to run all CI jobs on all PRs so we try to limit adding ciflow/rocm to all PRs, only the ones that are known to specifically target ROCm features or code paths.  When a PR doesn't fail the subset of ROCm tests, but breaks after merge (unfortunate, yes), we recommend adding ciflow/rocm to the reverted PR at that point.

**Pull request closed without merging**
